src/harness/fakes.ts

```diff
@@ -0,0 +1,376 @@
+/**
+ * Fake implementations of various compiler dependencies.
+ */
+namespace fakes {
+    const processExitSentinel = new Error("System exit");
+
+    export interface SystemOptions {
+        executingFilePath?: string;
+        newLine?: "\r\n" | "\n";
+        env?: Record<string, string>;
+    }
+
+    /**
+     * A fake `ts.System` that leverages a virtual file system.
+     */
+    export class System implements ts.System {
+        public readonly vfs: vfs.FileSystem;
+        public readonly args: string[] = [];
+        public readonly output: string[] = [];
+        public readonly newLine: string;
+        public readonly useCaseSensitiveFileNames: boolean;
+        public exitCode: number;
+
+        private readonly _executingFilePath: string | undefined;
+        private readonly _env: Record<string, string> | undefined;
+
+        constructor(vfs: vfs.FileSystem, { executingFilePath, newLine = "\r\n", env }: SystemOptions = {}) {

```

Comment:
Why do you have this another layer? Why not make vfs.FileSystem implement ts.System ?

Reply:
I'd rather keep them separate so that `FileSystem` is purely a file system abstraction. Breaking things down into multiple components is more maintainable and better delineates functionality. We already have too many monolithic dependencies in the compiler, and we don't always need a full `System` implementation (which is why we have so many independent `*Host` types).